src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -54,16 +49,34 @@ module.exports = function<T, P, I>(config : HostConfig<T, P, I>) : Reconciler {
 
   let nextUnitOfWork : ?Fiber = null;
 
-  function performUnitOfWork(unit : Fiber) : ?Fiber {
-    switch (unit.tag) {
-      case FunctionalComponent:
-        return ReactFiberFunctionalComponent.performWork(unit);
-      case ClassComponent:
-        break;
-      case HostComponent:
-        break;
+  function completeUnitOfWork(unitOfWork : Fiber) : ?Fiber {
+    while (true) {
+      var next = completeWork(unitOfWork);
+      if (next) {
+        // If completing this work spawned new work, do that next.
+        return next;
+      } else if (unitOfWork.sibling) {
+        // If there is more work to do in this parent, do that next.
+        return unitOfWork.sibling;
+      } else if (unitOfWork.parent) {
+        // If there's no more work in this parent. Complete the parent.
```

Comment:
This is how completion works by walking up the parent.

I was, however, considering using an explicit stack object that contains the path to the parent currently executing since that can be recreated whenever you start back from the top. That saves us having to store the parent pointer on the fiber itself.

I figured we might need the parent point for other things like finding effects, or dispatching events, but we could also do that using something like "host parent" or "event parent" like we do in the stack reconciler.